File: Leetcode Challenge/February/cpp/Minimize Deviation in Array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        priority_queue<int> maxpq; //max heap
        int m = INT_MAX, diff = INT_MAX;
        for(auto i : nums){
            if(i%2 != 0) i*=2;
            m = min(m, i);
            maxpq.push(i);
        }
        
        while(maxpq.top()%2 == 0) {
            int mx = maxpq.top();
            maxpq.pop();
            diff = min(diff, mx - m);
            m = min(m, mx/2);
            maxpq.push(mx/2);
        }
    
        return min(diff, maxpq.top() - m);
    }
};

File: Leetcode Challenge/February/java/Minimize Deviation in Array.java
class Solution {
    public int minimumDeviation(int[] nums) {
        PriorityQueue<Integer> pq=new PriorityQueue<>((a,b)->b-a); 
        
        int min=Integer.MAX_VALUE; 
        for(int i:nums){ 
            if(i%2==1) i*=2;
            min=Math.min(min,i); 
            pq.add(i);
        }
        
        int diff=Integer.MAX_VALUE; 
        while(pq.peek()%2==0){ 
            int max=pq.remove();
            diff=Math.min(diff,max-min);
            min=Math.min(max/2,min);
            pq.add(max/2); 
        }
        
        return Math.min(diff,pq.peek()-min);
    }
}
